File: julia/abc200_209/abc201_E.jl

using Random
infile = stdin
## Type Shortcuts (to save my wrists and fingers :))
const I = Int64; const VI = Vector{I}; const SI = Set{I}; const PI = NTuple{2,I};
const TI = NTuple{3,I}; const QI = NTuple{4,I}; const VPI = Vector{PI}; const SPI = Set{PI}
const VC = Vector{Char}; const VS = Vector{String}; VB = Vector{Bool}; VVI = Vector{Vector{Int64}}
const F = Float64; const VF = Vector{F}; const PF = NTuple{2,F}

gs()::String = rstrip(readline(infile))
gi()::Int64 = parse(Int64, gs())
gf()::Float64 = parse(Float64,gs())
gss()::Vector{String} = split(gs())
gis()::Vector{Int64} = [parse(Int64,x) for x in gss()]
gfs()::Vector{Float64} = [parse(Float64,x) for x in gss()]

function solve(N::I,u::VI,v::VI,w::VI)
    adj::Vector{VPI} = [VPI() for _ in 1:N ]
    for i in 1:N-1
        push!(adj[u[i]],(v[i],w[i]))
        push!(adj[v[i]],(u[i],w[i]))
    end
    dist::VI = fill(0,N)
    q::VPI = [(1,-1)]
    while !isempty(q)
        (n::I,p::I) = popfirst!(q)
        for (c::I,ww::I) in adj[n]
            if c == p; continue; end
            dist[c] = dist[n] ⊻ ww
            push!(q,(c,n))
        end
    end
    ones::VI = fill(0,60)
    zeros::VI = fill(0,60)
    for i in 1:N
        mask::I = 1
        d::I = dist[i]
        for j in 1:60
            if d & mask != 0; ones[j] += 1; else zeros[j] += 1; end
            mask <<= 1
        end
    end
    ans::I = 0
    mm::I = 10^9+7
    for i in 1:60
        adder::I = (1 << (i-1)) % mm * ones[i] % mm * zeros[i] % mm
        ans = (ans + adder) % mm
    end
    return ans
end

function main(infn="")
    global infile
    infile = (infn != "") ? open(infn,"r") : length(ARGS) > 0 ? open(ARGS[1],"r") : stdin
    N::I = gi()
    u::VI = fill(0,N-1)
    v::VI = fill(0,N-1)
    w::VI = fill(0,N-1)
    for i in 1:N-1; u[i],v[i],w[i] = gis(); end
    ans = solve(N,u,v,w)
    println(ans)
end

Random.seed!(8675309)
main()

#using Profile, StatProfilerHTML
#Profile.clear()
#@profile main("B.in")
#Profile.clear()
#@profilehtml main("B.in")

File: python/abc200_209/abc201_E.py

import sys
infile = sys.stdin.buffer
import collections

def gs()  : return infile.readline().rstrip()
def gi()  : return int(gs())
def gf()  : return float(gs())
def gss() : return gs().split()
def gis() : return [int(x) for x in gss()]
def gfs() : return [float(x) for x in gss()]

def main(infn="") :
    global infile
    infile = open(infn,"r") if infn else open(sys.argv[1],"r") if len(sys.argv) > 1 else sys.stdin
    ## Trick1 -- xor of weights of path from a to b is same as xor of weights of path from a to root and from b to root
    ## Trick2 -- use bitwise math to do the sum part with binomial coefficients
    N = gi()
    u = [0] * (N-1)
    v = [0] * (N-1)
    w = [0] * (N-1)
    for i in range(N-1) : u[i],v[i],w[i] = gis()
    gr = [[] for i in range(N)]
    for i in range(N-1) :
        gr[u[i]-1].append((v[i]-1,w[i]))
        gr[v[i]-1].append((u[i]-1,w[i]))
    dist = [-1] * N

    ## Use BFS, since 2*10**5 DFS is unreliable in python
    q = collections.deque()
    q.append(0)
    dist[0] = 0
    while q :
        n = q.popleft()
        for (c,wc) in gr[n] :
            if dist[c] >= 0 : continue ## Parent check
            dist[c] = dist[n] ^ wc
            q.append(c)
    ans = 0
    mm = 10**9+7
    for i in range(62) :
        numzeros = 0
        tag = 1<<i
        for j in range(N) :
            if dist[j] & (1<<i) == 0 : numzeros += 1
        numones = N - numzeros
        adder = (1 << i) * numones % mm * numzeros % mm
        ans = (ans + adder) % mm
    sys.stdout.write(str(ans)+'\n')

if __name__ == '__main__' :
    main()
    sys.stdout.flush()

